snow/consensus/snowman/poll/prefix.go

```diff
@@ -0,0 +1,186 @@
+// Copyright (C) 2019-2024, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package poll
+
+import (
+	"fmt"
+
+	"github.com/ava-labs/avalanchego/ids"
+)
+
+// longestSharedPrefixes creates a prefixGroup that is the root of a graph
+// of prefixGroup vertices.
+// When iterating the graph, each prefixGroup vertex represents a shared bit prefix
+// of IDs, and the members field contains all IDs from the given idList for which their bit prefix
+// matches the prefix field.
+func longestSharedPrefixes(idList []ids.ID) *prefixGroup {
+	originPG := &prefixGroup{members: idList}
+
+	pgs := make([]*prefixGroup, 0, len(idList))
+	pgs = append(pgs, originPG)
+
+	// Try to split each prefix group.
+	// Continue until all prefix groups cannot be split anymore.
+	for {
+		var couldSplit bool
+		for _, pg := range pgs {
+			if !pg.canSplit() {
+				continue
+			}
+
+			couldSplit = true
+
+			pg0, pg1 := pg.split()
+
+			if pg0 != nil && pg1 != nil {
+				pgs = append(pgs, pg0)
+				pgs = append(pgs, pg1)
+				continue
+			}
+
+			// Else, there is no bifurcation,
+			// so swallow up your descendant
+			descendant := determineDescendant(pg)
+
+			// Backup the parent
+			parent := pg.parent
+			// Become your descendant, but keep your original parent
+			*pg = *descendant
+			pg.parent = parent
+		}
+
+		if !couldSplit {
+			break
+		}
+	}
+
+	return originPG
+}
+
+func determineDescendant(pg *prefixGroup) *prefixGroup {
```

Comment:
It will be cleaner code wise, but I want to emphasize the different semantics here - the case where both are not nil is a bifurcation, and the complement of that is that we extend the current prefix as you correctly pointed out. These two cases are not the same and therefore to make it clearer in the code, I did not put them in a single switch statement.

If you feel strongly about it, I can collapse them in a switch statement, but I thought it would be more clear to the reader this way.